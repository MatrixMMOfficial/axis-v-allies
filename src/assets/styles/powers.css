li {
  list-style: none;
}
#powers {
  margin-bottom: 0;
  display: flex;
  font-family: 'Diplomata SC';
  -webkit-margin-before: 0;
  -webkit-padding-start: 0;
  background: #000;
}

#powers li {
  width: 14%;
  border: 3px solid black;
  font-size: 36px;
  opacity: 0.8;
}
#powers li.active {
  opacity: 1;
  color: #fefefe;
}

li div.flag {
  vertical-align: top;
}

#powers li img {
  vertical-align: top;
  display: inline-block;
}

li img.round {
  border-radius: 50%;
  height: 36px;
  width: 36px;
  background: #fff;
}

li.china img.round {
  background: #e01500;
}

li div {
  display: inline-block;
  padding: 5px 0 0 5px;
}

li div h1 {
  margin: 0;
  font-size: 0.45em;
  line-height: 0.9;
}

li div h2 {
  margin: 0;
  font-size: 0.4em;
}
