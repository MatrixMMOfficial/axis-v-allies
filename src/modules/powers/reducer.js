import initialPowers from '../../config/initialPowers';
import { omit } from 'ramda';

const updateObject = (object, newValues) => {
  return Object.assign({}, object, newValues)
}

const updateCurrentPower = (powers, updateCallback, callbackArg) => {
  return powers.map(power => power.current ? updateCallback(power, callbackArg) : power)
}

const spendIPCs = (power, amount) => {
  return updateObject(power, { ipc: power.ipc - amount })
}

const gainIPCs = (power, amount) => {
  return updateObject(power, { ipc: power.ipc +  amount })
}

const powers = (state = initialPowers, action) => {
  switch (action.type) {
    case 'NEXT_TURN':
      const currentPower = state.find(power => power.current)
      const nextPowerIndex = currentPower.name === 'China' ? 0 : state.indexOf(currentPower) + 1
      return state.map((power, n) => {
        if(power.current) {
          return omit('current', power) 
        } else if (n === nextPowerIndex) {
          return {...power, current: true}
        } else {
          return power
        }
      })
    case 'DEVELOP_TECH':
      const assignTech = (power) => {
        return updateObject(power, { tech: power.tech.concat(action.tech) })
      }
      return updateCurrentPower(state, assignTech)
    case 'ATTEMPT_RESEARCH':
      return updateCurrentPower(state, spendIPCs, action.cost)
    case 'INCREMENT_PURCHASE':
      return updateCurrentPower(state, spendIPCs, action.unit.cost)
    case 'DECREMENT_PURCHASE':
      return updateCurrentPower(state, gainIPCs, action.unit.cost)
    default:
      return state
  }
}

export default powers

