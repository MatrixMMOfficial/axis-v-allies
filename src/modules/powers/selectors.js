import { createSelector } from 'reselect'
import { getCurrentPower, getPowers } from '../../selectors/getCurrentPower'
import { territoriesOwnedBy, calculateNPL } from '../income'
export { getCurrentPower, getPowers }

const nplFor = (power, territories) => {
  return power === 'China' ? territories.length : calculateNPL(territories)
}

const nplAsObject = (powers, territories) => {
  return powers.reduce((object, { name }) => {
    object[name] = nplFor(name, territories[name])
    return object
  }, {})
}

export const nplByPower = createSelector(
  getPowers,
  territoriesOwnedBy,
  nplAsObject
)

