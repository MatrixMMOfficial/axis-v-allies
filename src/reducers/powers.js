import initialPowers from '../config/initialPowers'
import {
  ATTEMPT_RESEARCH,
  DEVELOP_TECH,
  INCREMENT_PURCHASE,
  DECREMENT_PURCHASE,
  VIEW_STRATEGIC_BOMBING_RESULTS,
  ASSESS_ROCKET_DAMAGE
} from '../actions'

const updateCurrentPower = (powers, cPI, updateCallback, callbackArg) => (
  powers.map((power, i) => i === cPI ? updateCallback(power, callbackArg) : power)
)

const spendIPCs = (power, amount) => (
  { ...power, ipc: power.ipc - amount }
)

const gainIPCs = (power, amount) => (
  { ...power, ipc: power.ipc +  amount }
)

const powers = (state = initialPowers, action) => {
  const { currentPowerIndex } = action //state
  switch (action.type) {
    case '@@router/LOCATION_CHANGE': 
      const { pathname } = action.payload.location
      if (pathname === '/confirm-finish') {
        //TODO: how to get CPI in here?
        return updateCurrentPower(state, currentPowerIndex, gainIPCs, state.currentPowerIncome)
      } else {
        return state
      }
    case VIEW_STRATEGIC_BOMBING_RESULTS: {
      return state.map(power => {
        if (power.name === action.power) {
          return spendIPCs(power, action.damage)
        } else {
          return power
        }
      })
    }
    case ASSESS_ROCKET_DAMAGE: {
      return state.map(power => {
        if (action.damages[power.name]) {
          return spendIPCs(power, action.damages[power.name])
        } else {
          return power
        }
      })
    }
    case DEVELOP_TECH:
      const assignTech = (power) => {
        return { ...power, tech: power.tech.concat(action.tech) }
      }
      return updateCurrentPower(state, currentPowerIndex, assignTech)
    case ATTEMPT_RESEARCH:
      return updateCurrentPower(state, currentPowerIndex, spendIPCs, action.cost)
    case INCREMENT_PURCHASE:
      return updateCurrentPower(state, currentPowerIndex, spendIPCs, action.unit.cost)
    case DECREMENT_PURCHASE:
      return updateCurrentPower(state, currentPowerIndex, gainIPCs, action.unit.cost)
    default:
      return state
  }
}

export default powers


