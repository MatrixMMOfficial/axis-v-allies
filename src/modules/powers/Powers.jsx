import React from 'react'
import classNames from 'classnames'
import { flags } from '../../config/initialPowers'

import '../../assets/styles/powers.css'

const Powers = ({ powers, currentPower, npl }) => {
  return (
    <ul id="powers">
      {powers.map((power, index) => {
        return <Power
          power={power}
          currentPower={currentPower}
          npl={npl[power.name]}
          key={index} />
        }
      )}
    </ul>
  )
}

const normalPowerIPC = (power, npl) => {
  return (
    <div>
      <h1>{power.name}</h1>
      <h2>{power.ipc} I.P.C.s</h2>
      <h2>{npl} N.P.L.</h2>
    </div>
  )
}

const chinaIPC = (npl) => {
  return (
    <div title={`${npl} Territories`}>
      <h1>China</h1>
      <h2>{npl}</h2>
    </div>
  )
}

const renderIPC = (power, npl) => {
  return power.name === 'China' ? chinaIPC(npl) : normalPowerIPC(power, npl)
}

const Power = ({power, currentPower, npl}) => {
  const powerName = power.name.toLowerCase()
  const liClass = classNames(powerName, { active: currentPower === power })
  return (
    <li className={liClass}>
      <div className="flag">
        <img
          className="round"
          src={flags[powerName]}
          alt={power.name} />
      </div>
      {renderIPC(power, npl)}
    </li>
  )
}
export default Powers
