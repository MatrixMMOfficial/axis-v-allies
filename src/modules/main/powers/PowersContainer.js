// @flow
import { connect } from 'react-redux'
import Powers from './Powers'
import { getPowers, getCurrentPower, nplByPower } from './selectors'

const mapStateToProps = (state) => {
  return {
    powers: getPowers(state),
    currentPower: getCurrentPower(state),   
    npl: nplByPower(state)
  }
}

const PowersContainer = connect(mapStateToProps)(Powers)

export default PowersContainer

