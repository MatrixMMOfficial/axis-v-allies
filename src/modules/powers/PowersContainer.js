import { connect } from 'react-redux'
import Powers from './Powers'
import { getCurrentPower, nplByPower } from './selectors'

const mapStateToProps = (state) => {
  return {
    powers: state.powers,
    currentPower: getCurrentPower(state),   
    npl: nplByPower(state)
  }
}

const PowersContainer = connect(mapStateToProps)(Powers)

export default PowersContainer

