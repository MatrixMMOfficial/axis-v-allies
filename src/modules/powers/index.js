import PowersContainer from './PowersContainer';

export { PowersContainer }
