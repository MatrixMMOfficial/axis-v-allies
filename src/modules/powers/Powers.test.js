'use strict'

import React from 'react'
import Powers from './Powers'
import renderer from 'react-test-renderer'
import powers from '../../config/initialPowers'

const npl = powers.reduce((obj, power) => {
  obj[power.name] = power.ipc
  return obj
}, {})

it('renders correctly', () => {
  const tree = renderer.create(
    <Powers powers={powers} npl={npl} currentPower={'Germany'} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

