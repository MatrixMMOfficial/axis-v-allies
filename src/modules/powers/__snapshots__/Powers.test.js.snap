// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ul
  id="powers"
>
  <li
    className="germany"
  >
    <div
      className="flag"
    >
      <img
        alt="Germany"
        className="round"
        src="germany.svg"
      />
    </div>
    <div>
      <h1>
        Germany
      </h1>
      <h2>
        56
         I.P.C.s
      </h2>
      <h2>
        56
         N.P.L.
      </h2>
    </div>
  </li>
  <li
    className="ussr"
  >
    <div
      className="flag"
    >
      <img
        alt="USSR"
        className="round"
        src="ussr.svg"
      />
    </div>
    <div>
      <h1>
        USSR
      </h1>
      <h2>
        46
         I.P.C.s
      </h2>
      <h2>
        46
         N.P.L.
      </h2>
    </div>
  </li>
  <li
    className="japan"
  >
    <div
      className="flag"
    >
      <img
        alt="Japan"
        className="round"
        src="japan.svg"
      />
    </div>
    <div>
      <h1>
        Japan
      </h1>
      <h2>
        41
         I.P.C.s
      </h2>
      <h2>
        41
         N.P.L.
      </h2>
    </div>
  </li>
  <li
    className="uk"
  >
    <div
      className="flag"
    >
      <img
        alt="UK"
        className="round"
        src="uk.svg"
      />
    </div>
    <div>
      <h1>
        UK
      </h1>
      <h2>
        70
         I.P.C.s
      </h2>
      <h2>
        70
         N.P.L.
      </h2>
    </div>
  </li>
  <li
    className="italy"
  >
    <div
      className="flag"
    >
      <img
        alt="Italy"
        className="round"
        src="italy.svg"
      />
    </div>
    <div>
      <h1>
        Italy
      </h1>
      <h2>
        27
         I.P.C.s
      </h2>
      <h2>
        27
         N.P.L.
      </h2>
    </div>
  </li>
  <li
    className="us"
  >
    <div
      className="flag"
    >
      <img
        alt="US"
        className="round"
        src="us.svg"
      />
    </div>
    <div>
      <h1>
        US
      </h1>
      <h2>
        67
         I.P.C.s
      </h2>
      <h2>
        67
         N.P.L.
      </h2>
    </div>
  </li>
  <li
    className="china"
  >
    <div
      className="flag"
    >
      <img
        alt="China"
        className="round"
        src="china.svg"
      />
    </div>
    <div
      title="undefined Territories"
    >
      <h1>
        China
      </h1>
      <h2 />
    </div>
  </li>
</ul>
`;
